### PR TITLE
DBZ-683 Snapshot supports storing only whitelisted tables setting

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.jboss.org/browse/DBZ-683

### CODE REVIEW

**Comment by gunnarmorling** [comment]
@jpechane, could you rebase this? There are some conflicts due to the parser removal.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java [hunk 1]
```diff
@@ -302,12 +302,12 @@ protected void execute() {
                         mysql.query(sql.get(), rs -> {
                             while (rs.next() && isRunning()) {
                                 TableId id = new TableId(dbName, null, rs.getString(1));
-                                if (createTableFilters.tableFilter().test(id)) {
+                                if ((createTableFilters == filters && shouldRecordTableSchema(schema, filters, id)) || createTableFilters.tableFilter().test(id)) {
```
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java]: Would be good to store the result of the `shouldRecordTableSchema()` call as it's done below again.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/antlr/listener/UseStatementParserListener.java [hunk 2]
```diff
@@ -35,6 +35,9 @@ public void enterUseStatement(MySqlParser.UseStatementContext ctx) {
         // right for the current database.
         String charsetForDb = parser.charsetNameForDatabase().get(dbName);
         parser.systemVariables().setVariable(MySqlSystemVariables.MySqlScope.SESSION, MySqlSystemVariables.CHARSET_NAME_DATABASE, charsetForDb);
+
+        // Signal that the variable was set ...
+        parser.signalUseDatabase(ctx);
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/antlr/listener/UseStatementParserListener.java]: Interesting, why is that needed now?

File: debezium-core/src/main/java/io/debezium/relational/ddl/DdlChanges.java [hunk 3]
```diff
@@ -162,4 +166,15 @@ public String toString() {
     public static interface DatabaseStatementStringConsumer {
         void consume(String databaseName, String ddlStatements);
     }
+
+    public boolean anyMatch(Predicate<String> databaseFilter, Predicate<TableId> tableFilter) {
```
[the remark this answers is not in the document]
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/relational/ddl/DdlChanges.java]: @gunnarmorling Javadoc added

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlSchema.java [hunk 4]
```diff
@@ -317,7 +317,7 @@ public boolean applyDdl(SourceInfo source, String databaseName, String ddlStatem
             // Note that, unlike with the DB history topic, we don't filter out non-whitelisted tables here
```
[the remark this answers is not in the document]
Comment (jpechane) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlSchema.java]: @gunnarmorling comment removed

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpechane) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java]: @gunnarmorling The `==` is intentional. I've added the comment to it.

File: debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySqlConnectorIT.java [hunk 5]
```diff
@@ -1054,6 +1058,57 @@ public void shouldIgnoreCreateIndexForNonCapturedTablesNotStoredInHistory() thro
...
+            }
+        }
+
+        // Start the connector ...
+        start(MySqlConnector.class, config);
+
+        // Consume the first records due to startup and initialization of the database ...
+        // Testing.Print.enable();
+        // Four tables in database, only two are whitelisted
+        SourceRecords records = consumeRecordsByTopic(1 + 3 + 2 * 4 + 3 + 2);
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySqlConnectorIT.java]: So I think these additions help a *little bit*, but still it's not quite clear what the expected value represents. Can we clarify this in terms of docs somehow? I don't have a good answer, but the current form is really hard to grasp.

Comment (jpechane) [reply debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySqlConnectorIT.java]: @gunnarmorling Updated

**Comment by gunnarmorling** [comment]
@jpechane, there's a test failure; I suppose caused by concurrent related PR being merged in the meantime?

**Comment by jpechane** [comment]
@gunnarmorling Rebased, rebased, rebased, pushed and finally passing :-)

**Comment by gunnarmorling** [comment]
Ok, excellent!
